boudicca.base/common-model/src/main/kotlin/base/boudicca/model/structured/AbstractKey.kt

```diff
@@ -0,0 +1,97 @@
+package base.boudicca.model.structured
+
+import kotlin.math.min
+
+/**
+ * represents a parsed Key of a Key-Value pair which consists of the name and all the variants (which are sorted canonically)
+ */
+abstract class AbstractKey<T : AbstractKey<T>>(val name: String, variants: List<Variant> = emptyList()) :
+    Comparable<T> {
+
+    val variants = variants.sorted()
+
+    init {
+        internalValidate()
+    }
+
+    /**
+     * get the key string representation of this key for using in an unstructured/serialized event
+     */
+    fun toKeyString(): String {
+        if (variants.isEmpty()) {
+            return name
+        }
+        return name + ":" + variants.joinToString(":") { it.toKeyString() }
+    }
+
+    companion object {
+        val COMPARATOR = compareBy<AbstractKey<*>> { it.name }.thenComparing { o1, o2 ->
+            for (i in 0..<min(o1.variants.size, o2.variants.size)) {
```

Comment:
you don't need a for loop here as progressions are iterable
```suggestion
           (0..min(o1.variants.size, o2.variants.size)).forEach { i ->
```